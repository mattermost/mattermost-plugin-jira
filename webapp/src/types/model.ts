export type ReactSelectOption = {
    label: string | React.ReactElement;
    value: string;
};

export type AllowedValue = {
    id: string;
    value?: string;
    name?: string;
};

export type FieldSchema = {
    type: string;
    system?: string;
    custom?: JiraFieldCustomTypeEnums;
    customId?: number;
    items?: string;
};

export type BaseField = {
    key?: string;
    name: string;
    required: boolean;
    schema: FieldSchema;
    allowedValues?: AllowedValue[];
};

export type SelectField = BaseField & {
    allowedValues: AllowedValue[];
};

export type StringArrayField = BaseField;

export type StringField = BaseField;

export type JiraField = SelectField | StringArrayField | StringField;

export type IssueTypeIdentifier = {id: string; name: string};

export type FilterField = {
    key: string;
    name: string;
    schema: FieldSchema;
    values?: ReactSelectOption[];
    userDefined?: boolean;
    issueTypes: IssueTypeIdentifier[];
};

export type IssueType = {
    id: string;
    name: string;
    fields: {[key: string]: JiraField};
    subtask: boolean;
}

export type JiraIssue = {
    id: string;
    key: string;
    name: string;
    fields: {[key: string]: JiraField};
}

export type Project = {
    key: string;
    issuetypes: IssueType[];
}

export type IssueTypeWithStatuses = {
    id: string;
    name: string;
    statuses: Status[];
}

export type IssueMetadata = {
    projects: Project[];
    issue_types: IssueTypeWithStatuses[];
}

export type Status = {
    id: string;
    name: string;
}

export type ProjectMetadata = {
    projects: ReactSelectOption[];
    issues_per_project: {[key: string]: ReactSelectOption[]};
    default_project_key?: string;
}

export enum JiraFieldTypeEnums {
    PROJECT = 'project',
    ISSUE_TYPE = 'issuetype',
    PRIORITY = 'priority',
    DESCRIPTION = 'description',
    SUMMARY = 'summary',
    LABELS = 'labels',
    ASSIGNEE = 'assignee',
    SECURITY = 'security',
    COMPONENTS = 'components',
    FIX_VERSIONS = 'fixVersions',
    AFFECTS_VERSIONS = 'versions',
}

export enum AvatarSize {
    XSMALL = '16x16',
    SMALL = '24x24',
    MEDIUM = '36x36',
    LARGE = '48x48',
}

export type JiraUser = {
    self: string;
    accountId: string;
    avatarUrls: {[size in AvatarSize]: string};
    displayName: string;
    active: boolean;
    timeZone: string;
}

export enum JiraFieldCustomTypeEnums {
    SPRINT = 'com.pyxis.greenhopper.jira:gh-sprint',
    EPIC_LINK = 'com.pyxis.greenhopper.jira:gh-epic-link',
    EPIC_NAME = 'com.pyxis.greenhopper.jira:gh-epic-label',
    RANK = 'com.pyxis.greenhopper.jira:gh-lexo-rank',

    CASCADING_SELECT = 'com.atlassian.jira.plugin.system.customfieldtypes:cascadingselect',
    TEXT_AREA = 'com.atlassian.jira.plugin.system.customfieldtypes:textarea',
    TEXT_FIELD = 'com.atlassian.jira.plugin.system.customfieldtypes:textfield',
    SELECT = 'com.atlassian.jira.plugin.system.customfieldtypes:select',
    PROJECT = 'com.atlassian.jira.plugin.system.customfieldtypes:project',

    MULTI_SELECT = 'com.atlassian.jira.plugin.system.customfieldtypes:multiselect',
    RADIO_BUTTONS = 'com.atlassian.jira.plugin.system.customfieldtypes:radiobuttons',
    MULTI_CHECKBOXES = 'com.atlassian.jira.plugin.system.customfieldtypes:multicheckboxes',
    URL_FIELD = 'com.atlassian.jira.plugin.system.customfieldtypes:url',

    LABELS = 'com.atlassian.jira.plugin.system.customfieldtypes:labels',
    USER_PICKER = 'com.atlassian.jira.plugin.system.customfieldtypes:userpicker',
}

export enum FilterFieldInclusion {
    INCLUDE_ANY = 'include_any',
    INCLUDE_ALL = 'include_all',
    EXCLUDE_ANY = 'exclude_any',
    EMPTY = 'empty',
}

export type FilterValue = {
    key: string;
    values: string[];
    inclusion: FilterFieldInclusion;
}

export type ChannelSubscriptionFilters = {
    projects: string[];
    events: string[];
    issue_types: string[];
    fields: FilterValue[];
};

export type ChannelSubscription = {
    id: string;
    channel_id: string;
    filters: ChannelSubscriptionFilters;
    name: string;
    instance_id: string;
}

export enum InstanceType {
    CLOUD = 'cloud',
    SERVER = 'server',
}
export type Instance = {
    alias?: string;
    instance_id: string;
    type: InstanceType;
}

export type GetConnectedResponse = {
    data: {
        can_connect: boolean;
        instances: Instance[];
        is_connected: boolean;
        user: {
            connected_instances: Instance[];
            default_instance_id?: string;
        };
    };
    error?: Error;
};

export type APIResponse<T> = {
    error?: Error;
    data: T;
};

export type CreateIssueRequest = {
    instance_id: string;
    required_fields_not_covered: string[][];
    post_id: string;
    current_team: string;
    channel_id: string;
    fields: {};
};

export type SearchIssueParams = {
    jql?: string;
    fields: string;
    q: string;
    instance_id: string;
};

export type AttachCommentRequest = {
    post_id: string;
    current_team: string;
    issueKey: string;
    instance_id: string;
};

export type AllProjectMetadata = {
    instance_id: string;
    metadata: ProjectMetadata;
}[];

export type CreateIssueFields = {
    description: string;
    summary?: string;
    project: {key: string};
    issuetype: {id: string};
} & {[key: string]: JiraField};
