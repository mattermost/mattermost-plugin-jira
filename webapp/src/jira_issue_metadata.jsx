// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export function getProjectValues(metadata) {
    if (!metadata) {
        return [];
    }

    return metadata.projects.map((p) => ({value: p.key, label: p.name}));
}

export function getIssueTypes(metadata, projectKey) {
    return metadata.projects.find((project) => project.key === projectKey).issuetypes.filter((i) => !i.subtask);
}

export function getIssueValues(metadata, projectKey) {
    if (!metadata || !projectKey) {
        return [];
    }

    return getIssueTypes(metadata, projectKey).map((issueType) => ({value: issueType.id, label: issueType.name}));
}

export function getFields(metadata, projectKey, issueTypeId) {
    if (!metadata || !projectKey || !issueTypeId) {
        return [];
    }

    return getIssueTypes(metadata, projectKey).find((issueType) => issueType.id === issueTypeId).fields;
}

export function getFieldValues(metadata, projectKey, issueTypeId) {
    const fieldsForIssue = getFields(metadata, projectKey, issueTypeId);
    const fieldIds = Object.keys(fieldsForIssue);
    return fieldIds.map((fieldId) => ({value: fieldId, label: fieldsForIssue[fieldId].name}));
}

