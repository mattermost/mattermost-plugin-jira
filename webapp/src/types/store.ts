// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {GlobalState as ReduxGlobalState} from 'mattermost-redux/types/store';

import type combinedReducers from '../reducers';

export type GlobalState = ReduxGlobalState & {
    'plugins-jira': PluginState
};

export type PluginState = ReturnType<typeof combinedReducers>

export const pluginStateKey = 'plugins-jira' as const;
