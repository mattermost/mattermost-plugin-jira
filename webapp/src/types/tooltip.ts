import {JiraUser} from './model';

export type TicketDetails = {
    assigneeName: string;
    assigneeAvatar: string;
    labels: string[];
    description: string;
    summary: string;
    ticketId: string;
    jiraIcon: string;
    versions: string;
    statusKey: string;
    issueIcon: string;
}

export type TicketData = {
    key: string;
    fields: TicketDataFields;
}

export type AvatarUrls = {
    '48x48': string;
}

export type TicketDataFields = {
    assignee: JiraUser | null;
    labels: string[];
    description: string;
    summary: string;
    project: {avatarUrls: AvatarUrls};
    versions: string[];
    status: {name: string};
    issuetype: {iconUrl: string};
}

export type IssueAction = {
    type: string;
    data: TicketData;
}
