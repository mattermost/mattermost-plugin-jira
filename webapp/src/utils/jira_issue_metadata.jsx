// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export function getProjectValues(metadata) {
    if (!metadata) {
        return [];
    }

    return metadata.projects;
}

export function getIssueTypes(metadata, projectKey) {
    const project = metadata.projects.find((proj) => proj.key === projectKey);
    if (!project) {
        return [];
    }
    return project.issuetypes.filter((i) => !i.subtask);
}

export function getIssueValues(metadata, projectKey) {
    if (!metadata || !projectKey) {
        return [];
    }

    return metadata.issues_per_project[projectKey];
}

export function getIssueValuesForMultipleProjects(metadata, projectKeys) {
    return projectKeys.map((project) =>
        getIssueValues(metadata, project)).flat().filter(Boolean).sort((a, b) => a.value - b.value).filter((ele, i, me) => i === 0 || ele.value !== me[i - 1].value);
}

export function getFields(metadata, projectKey, issueTypeId) {
    if (!metadata || !projectKey || !issueTypeId) {
        return [];
    }

    return getIssueTypes(metadata, projectKey).find((issueType) => issueType.id === issueTypeId).fields;
}

export function getCustomFieldValuesForProject(metadata, projectKey) {
    if (!metadata || !projectKey) {
        return [];
    }

    const issueTypes = getIssueTypes(metadata, projectKey);

    const customFieldHash = {};
    const fields = issueTypes.map((it) => Object.values(it.fields)).flat();
    fields.forEach((field) => {
        const key = field.key;
        if (key.startsWith('customfield_')) {
            customFieldHash[key] = field;
        }
    });

    return Object.values(customFieldHash).sort((a, b) => {
        if (a.name < b.name) {
            return -1;
        }
        if (a.name > b.name) {
            return 1;
        }
        return 0;
    }).map((field) => ({
        label: `Issue Updated: Custom - ${field.name}`,
        value: `event_updated_${field.key}`,
    }));
}

export function getFieldValues(metadata, projectKey, issueTypeId) {
    const fieldsForIssue = getFields(metadata, projectKey, issueTypeId);
    const fieldIds = Object.keys(fieldsForIssue);
    return fieldIds.map((fieldId) => ({value: fieldId, label: fieldsForIssue[fieldId].name}));
}
