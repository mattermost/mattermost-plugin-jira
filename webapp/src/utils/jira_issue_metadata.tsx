// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {
    ProjectMetadata,
    ReactSelectOption,
    IssueMetadata,
    IssueType,
    JiraField,
    FilterField,
    FilterValue,
    SelectField,
    StringArrayField,
    IssueTypeIdentifier,
    ChannelSubscriptionFilters,
    FilterFieldInclusion,
    JiraFieldCustomTypeEnums,
    JiraFieldTypeEnums,
    Status,
    IssueTypeWithStatuses,
} from 'types/model';

type FieldWithInfo = JiraField & {
    changeLogID: string;
    topLevelKey: string;
    validIssueTypes: IssueTypeIdentifier[];
    issueTypeMeta: IssueTypeIdentifier;
}

export const FIELD_KEY_STATUS = 'status';

// This is a replacement for the Array.flat() function which will be polyfilled by Babel
// in our 5.16 release. Remove this and replace with .flat() then.
const flatten = (arr: any[]) => {
    return arr.reduce((acc, val) => acc.concat(val), []);
};

function sortByName<T>(arr: (T & {name: string})[]): T[] {
    return arr.sort((a, b) => {
        if (a.name < b.name) {
            return -1;
        }
        if (a.name > b.name) {
            return 1;
        }
        return 0;
    });
}

export function getProjectValues(metadata: ProjectMetadata | null): ReactSelectOption[] {
    if (!metadata || !metadata.projects) {
        return [];
    }

    return metadata.projects;
}

export function getIssueTypes(metadata: IssueMetadata | null, projectKey: string | null): IssueType[] {
    if (!metadata || !metadata.projects) {
        return [];
    }

    const project = metadata.projects.find((proj) => proj.key === projectKey);
    if (!project) {
        return [];
    }
    return project.issuetypes.filter((i) => !i.subtask);
}

export function getIssueValues(metadata: ProjectMetadata, projectKey: string): ReactSelectOption[] {
    if (!metadata || !metadata.issues_per_project || !projectKey) {
        return [];
    }

    return metadata.issues_per_project[projectKey];
}

export function getIssueValuesForMultipleProjects(metadata: ProjectMetadata, projectKeys: string[]): ReactSelectOption[] {
    if (!metadata || !metadata.projects || !projectKeys) {
        return [];
    }

    const issueValues = flatten(projectKeys.map((project) => getIssueValues(metadata, project))).filter(Boolean);

    const issueTypeHash: {[key: string]: ReactSelectOption} = {};
    issueValues.forEach((issueType: ReactSelectOption) => {
        issueTypeHash[issueType.value] = issueType;
    });

    return Object.values(issueTypeHash);
}

export function getFields(metadata: IssueMetadata | null, projectKey: string | null, issueTypeId: string | null): {[key: string]: JiraField} {
    if (!metadata || !projectKey || !issueTypeId) {
        return {};
    }

    const issueType = getIssueTypes(metadata, projectKey).find((it) => it.id === issueTypeId);
    if (issueType) {
        return issueType.fields;
    }
    return {};
}

export function getConflictingFields(fields: FilterField[], chosenIssueTypes: string[], issueMetadata: IssueMetadata): {field: FilterField; issueTypes: IssueType[]}[] {
    const conflictingFields = [];

    for (const field of fields) {
        const conflictingIssueTypes = [];
        for (const issueTypeId of chosenIssueTypes) {
            const issueTypes = field.issueTypes;
            if (!issueTypes.find((it) => it.id === issueTypeId)) {
                const issueType = issueMetadata.projects[0].issuetypes.find((i) => i.id === issueTypeId) as IssueType;
                conflictingIssueTypes.push(issueType);
            }
        }
        if (conflictingIssueTypes.length) {
            conflictingFields.push({field, issueTypes: conflictingIssueTypes});
        }
    }
    return conflictingFields;
}

export function getCustomFieldsForProjects(metadata: IssueMetadata | null, projectKeys: string[]): FieldWithInfo[] {
    if (!metadata || !projectKeys || !projectKeys.length) {
        return [];
    }

    const issueTypes = flatten(projectKeys.map((key) => getIssueTypes(metadata, key))) as IssueType[];

    const customFieldHash: {[key: string]: FieldWithInfo} = {};
    const fields = flatten(issueTypes.map((issueType) =>
        Object.keys(issueType.fields).map((key) => ({
            ...issueType.fields[key],
            topLevelKey: key,
            issueTypeMeta: {
                id: issueType.id,
                name: issueType.name,
            },
        }))
    )).filter(Boolean) as FieldWithInfo[];

    for (const field of fields) {
        // Jira server webhook fields don't have keys
        // name is the most unique property available in that case
        const changeLogID = field.key || field.name;
        let current = customFieldHash[field.topLevelKey];
        if (!current) {
            current = {...field, changeLogID, key: field.key || field.topLevelKey, validIssueTypes: []};
        }
        current.validIssueTypes.push(field.issueTypeMeta);

        customFieldHash[field.topLevelKey] = current;
    }

    return sortByName(Object.values(customFieldHash));
}

const allowedTypes = [
    'priority',
    'securitylevel',
    'security',
];

const allowedArrayTypes = [
    'component',
    'option', // multiselect
    'string', // labels
    'version', // fix and affects versions
];

const avoidedCustomTypesForFilters: string[] = [
    JiraFieldCustomTypeEnums.SPRINT,
];

const acceptedCustomTypesForFilters: string[] = [
    JiraFieldCustomTypeEnums.EPIC_LINK,
    JiraFieldCustomTypeEnums.CASCADING_SELECT,
];

function isValidFieldForFilter(field: JiraField): boolean {
    const {custom, type, items} = field.schema;
    if (custom && avoidedCustomTypesForFilters.includes(custom)) {
        return false;
    }

    return allowedTypes.includes(type) || (custom && acceptedCustomTypesForFilters.includes(custom)) ||
    type === 'option' || // single select
    (type === 'array' && allowedArrayTypes.includes(items));
}

export function getCustomFieldFiltersForProjects(metadata: IssueMetadata | null, projectKeys: string[], issueTypes: string[]): FilterField[] {
    const fields = getCustomFieldsForProjects(metadata, projectKeys).filter(isValidFieldForFilter);
    const selectFields = fields.filter((field) => Boolean(field.allowedValues && field.allowedValues.length)) as (SelectField & FieldWithInfo)[];
    const populatedFields = selectFields.map((field) => {
        return {
            key: field.key,
            name: field.name,
            schema: field.schema,
            values: field.allowedValues.map((value) => ({
                label: value.name || value.value,
                value: value.id,
            })),
            issueTypes: field.validIssueTypes,
        } as FilterField;
    });

    const stringArrayFields = fields.filter((field) => field.schema.type === 'array' && field.schema.items === 'string' && !field.allowedValues) as (StringArrayField & FieldWithInfo)[];
    const userDefinedFields = stringArrayFields.map((field) => {
        return {
            key: field.key,
            name: field.name,
            schema: field.schema,
            userDefined: true,
            issueTypes: field.validIssueTypes,
        } as FilterField;
    });

    const result = populatedFields.concat(userDefinedFields);
    const epicLinkField = fields.find(isEpicLinkField);
    if (epicLinkField) {
        result.unshift({
            key: epicLinkField.key,
            name: epicLinkField.name,
            schema: epicLinkField.schema,
            values: [],
            issueTypes: epicLinkField.validIssueTypes,
        } as FilterField);
    }

    // Filtering out the statuses on the basis of selected issue types
    const issueTypesWithStatuses = metadata && metadata.issue_types;
    const keys = new Set<string>();
    const statuses: Status[] = [];

    if (issueTypesWithStatuses) {
        for (const element of issueTypesWithStatuses) {
            if (issueTypes.includes(element.id) || !issueTypes.length) {
                for (const status of element.statuses) {
                    if (!keys.has(status.id)) {
                        keys.add(status.id);
                        statuses.push(status);
                    }
                }
            }
        }
    }

    if (statuses.length) {
        result.push({
            key: FIELD_KEY_STATUS,
            name: 'Status',
            schema: {
                type: 'array',
            },
            values: statuses.map((value) => ({
                label: value.name,
                value: value.id,
            })),
            issueTypes: metadata && metadata.issue_types.map((type) => {
                return {
                    id: type.id,
                    name: type.name,
                };
            }),
        } as FilterField);
    }

    return sortByName(result);
}

const avoidedCustomTypesForEvents: string[] = [
    JiraFieldCustomTypeEnums.SPRINT,
    JiraFieldCustomTypeEnums.RANK,
];

function isValidFieldForEvents(field: JiraField): boolean {
    const {custom} = field.schema;
    if (!custom) {
        return false;
    }

    return !avoidedCustomTypesForEvents.includes(custom);
}

export function getCustomFieldValuesForEvents(metadata: IssueMetadata | null, projectKeys: string[]): ReactSelectOption[] {
    return getCustomFieldsForProjects(metadata, projectKeys).filter(isValidFieldForEvents).map((field) => ({
        label: `Issue Updated: Custom - ${field.name}`,
        value: `event_updated_${field.changeLogID}`,
    }));
}

export function getFieldValues(metadata: IssueMetadata, projectKey: string, issueTypeId: string): ReactSelectOption[] {
    const fieldsForIssue = getFields(metadata, projectKey, issueTypeId);
    const fieldIds = Object.keys(fieldsForIssue);
    return fieldIds.map((fieldId) => ({value: fieldId, label: fieldsForIssue[fieldId].name}));
}

export function isEpicNameField(field: JiraField | FilterField): boolean {
    return field.schema && field.schema.custom === JiraFieldCustomTypeEnums.EPIC_NAME;
}

export function isEpicLinkField(field: JiraField | FilterField): boolean {
    return field.schema && field.schema.custom === JiraFieldCustomTypeEnums.EPIC_LINK;
}

export function isLabelField(field: JiraField | FilterField): boolean {
    return field.schema.system === 'labels' || field.schema.custom === 'com.atlassian.jira.plugin.system.customfieldtypes:labels';
}

export function isEpicIssueType(issueType: IssueType): boolean {
    return issueType.name === 'Epic';
}

export function isMultiSelectField(field: FilterField): boolean {
    return field.schema.type === 'array';
}

export function isTextField(field: JiraField | FilterField): boolean {
    return field.schema.type === 'string';
}

export function isSecurityLevelField(field: JiraField | FilterField): boolean {
    return field.schema.type === 'securitylevel';
}

export function filterValueIsSecurityField(value: FilterValue): boolean {
    return value.key === JiraFieldTypeEnums.SECURITY;
}

// Some Jira fields have special names for JQL
function getFieldNameForJQL(field: FilterField) {
    switch (field.key) {
    case 'fixVersions':
        return 'fixVersion';
    case 'versions':
        return 'affectedVersion';
    default:
        return field.name;
    }
}

function quoteGuard(s: string) {
    if (s && s.includes(' ')) {
        return `"${s}"`;
    }

    return s;
}

export function generateJQLStringFromSubscriptionFilters(issueMetadata: IssueMetadata, fields: FilterField[], filters: ChannelSubscriptionFilters, securityLevelEmptyForJiraSubscriptions: boolean) {
    const projectJQL = `Project = ${quoteGuard(filters.projects[0]) || '?'}`;

    let issueTypeValueString = '?';
    if (filters.issue_types.length) {
        const issueTypeNames = filters.issue_types.map((issueTypeId) => {
            const issueType = issueMetadata.projects[0].issuetypes.find((it) => it.id === issueTypeId);
            if (!issueType) {
                return issueTypeId;
            }

            return `${quoteGuard(issueType.name)}`;
        });
        issueTypeValueString = `(${issueTypeNames.join(', ')})`;
    }
    const issueTypesJQL = `IssueType IN ${issueTypeValueString}`;

    let filterFieldsJQL = filters.fields.map(({key, inclusion, values}): string => {
        const field = fields.find((f) => f.key === key);
        if (!field) {
            // broken filter
            return `(cannot find field ${key})`;
        }

        const fieldName = getFieldNameForJQL(field);

        if (inclusion === FilterFieldInclusion.EMPTY) {
            return `${quoteGuard(fieldName)} IS EMPTY`;
        }

        const inclusionString = inclusion === FilterFieldInclusion.EXCLUDE_ANY ? 'NOT IN' : 'IN';
        if (!values.length) {
            return `${quoteGuard(fieldName)} ${inclusionString} ?`;
        }

        const chosenValueLabels = values.map((value) => {
            if (!(field.values && field.values.length)) {
                return value;
            }

            const found = field.values.find((v) => v.value === value);
            if (!found) {
                return value;
            }

            return found.label;
        });

        if (inclusion === FilterFieldInclusion.INCLUDE_ALL && values.length > 1) {
            const clauses = chosenValueLabels.map((v) => `${quoteGuard(fieldName)} IN (${quoteGuard(v)})`);
            return `(${clauses.join(' AND ')})`;
        }

        const joinedValues = chosenValueLabels.map((v) => `${quoteGuard(v)}`).join(', ');
        const valueString = `(${joinedValues})`;
        return `${quoteGuard(fieldName)} ${inclusionString} ${valueString}`;
    }).join(' AND ');

    const shouldShowEmptySecurityLevel = securityLevelEmptyForJiraSubscriptions && !filters.fields.some(filterValueIsSecurityField);
    if (shouldShowEmptySecurityLevel) {
        if (filterFieldsJQL.length) {
            filterFieldsJQL += ' AND ';
        }
        filterFieldsJQL += '"Security Level" IS EMPTY';
    }

    return [projectJQL, issueTypesJQL, filterFieldsJQL].filter(Boolean).join(' AND ');
}
