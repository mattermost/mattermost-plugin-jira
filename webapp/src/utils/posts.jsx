// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

// TODO: change to use the isCombinedUserActivityPost in 'mattermost-redux/utils/post_list' when we upgrade to 5.12
export const isCombinedUserActivityPost = (id) => {
    return (/^user-activity-(?:[^_]+_)*[^_]+$/).test(id);
};

