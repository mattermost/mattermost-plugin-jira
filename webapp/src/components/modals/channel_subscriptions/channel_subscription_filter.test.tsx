// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import JiraEpicSelector from 'components/data_selectors/jira_epic_selector';

import issueMetadata from 'testdata/cloud-get-create-issue-metadata-for-project.json';

import {FilterFieldInclusion, IssueMetadata, FilterField} from 'types/model';
import {getCustomFieldFiltersForProjects, isEpicLinkField} from 'utils/jira_issue_metadata';

import ChannelSubscriptionFilter, {Props} from './channel_subscription_filter';

describe('components/ChannelSubscriptionFilter', () => {
    const fields = getCustomFieldFiltersForProjects(issueMetadata, [issueMetadata.projects[0].key], []);
    const baseProps: Props = {
        theme: {},
        fields,
        field: fields.find((f) => f.key === 'priority') as FilterField,
        value: {
            key: 'priority',
            inclusion: FilterFieldInclusion.INCLUDE_ANY,
            values: ['1'],
        },
        chosenIssueTypes: [],
        issueMetadata: issueMetadata as IssueMetadata,
        addValidate: jest.fn(),
        removeValidate: jest.fn(),
        onChange: jest.fn(),
        removeFilter: jest.fn(),
        instanceID: 'https://something.atlassian.net',
        securityLevelEmptyForJiraSubscriptions: true,
    };

    test('should match snapshot', () => {
        const props = {...baseProps, issueMetadata: {}};
        const wrapper = shallow<ChannelSubscriptionFilter>(
            <ChannelSubscriptionFilter {...props}/>
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('should render JiraEpicSelector when Epic Link field is selected', () => {
        const props = {...baseProps};
        const wrapper = shallow<ChannelSubscriptionFilter>(
            <ChannelSubscriptionFilter {...props}/>
        );

        expect(wrapper.find(JiraEpicSelector).length).toBe(0);

        wrapper.setProps({
            ...props,
            field: fields.find(isEpicLinkField) as FilterField,
        });

        expect(wrapper.find(JiraEpicSelector).length).toBe(1);
    });

    test('should render correct inclusion captions for different include choices', () => {
        const props = {...baseProps};

        const wrapper = shallow<ChannelSubscriptionFilter>(
            <ChannelSubscriptionFilter {...props}/>
        );

        const select = wrapper.find('ReactSelectSetting[name="inclusion"]');
        const func = select.props().formatOptionLabel;

        const tests = [
            ['include_any', 'Includes either of the values (or)'],
            ['include_all', 'Includes all of the values (and)'],
            ['exclude_any', 'Excludes all of the values'],
            ['empty', 'Includes when the value is empty'],
        ];

        // Select dropdown is open
        for (const t of tests) {
            const element = func({value: t[0]}, {});
            const wrapper2 = shallow(element);
            expect(wrapper2.text()).toEqual(t[1]);
        }

        // Select dropdown is closed
        const result = func({value: 'include_any', label: 'Some Option Label'}, {context: 'value'});
        expect(result).toEqual('Some Option Label');
    });

    test('checkFieldConflictError should return an error string when there is a conflict', () => {
        const props = {
            ...baseProps,
            chosenIssueTypes: ['10002'],
            field: {
                ...baseProps.field,
                issueTypes: [{id: '10002', name: 'Task'}],
            },
        };
        const wrapper = shallow<ChannelSubscriptionFilter>(
            <ChannelSubscriptionFilter {...props}/>
        );

        let result;
        result = wrapper.instance().checkFieldConflictError();
        expect(result).toBeNull();

        wrapper.setProps({
            ...props,
            chosenIssueTypes: ['10002'],
            field: {
                ...props.field,
                name: 'FieldName',
                issueTypes: [{id: '10003', name: 'Task'}],
            },
        });

        result = wrapper.instance().checkFieldConflictError();
        expect(result).toEqual('FieldName does not exist for issue type(s): Task.');
    });

    test('checkInclusionError should return an error string when there is an invalid inclusion value', () => {
        const props: Props = {
            ...baseProps,
            field: {
                ...baseProps.field,
                schema: {
                    ...baseProps.field.schema,
                    type: 'securitylevel',
                },
            },
        };
        const wrapper = shallow<ChannelSubscriptionFilter>(
            <ChannelSubscriptionFilter {...props}/>
        );

        let isValid;
        isValid = wrapper.instance().isValid();
        expect(isValid).toBe(true);

        wrapper.setProps({
            ...props,
            value: {
                inclusion: FilterFieldInclusion.EMPTY,
                key: 'securitylevel',
                values: [],
            },
        });

        isValid = wrapper.instance().isValid();
        expect(isValid).toBe(true);

        wrapper.setProps({
            ...props,
            value: {
                inclusion: FilterFieldInclusion.INCLUDE_ANY,
                key: 'securitylevel',
                values: [],
            },
        });

        isValid = wrapper.instance().isValid();
        expect(isValid).toBe(true);

        wrapper.setProps({
            ...props,
            value: {
                inclusion: FilterFieldInclusion.EXCLUDE_ANY,
                key: 'securitylevel',
                values: [],
            },
        });

        isValid = wrapper.instance().isValid();
        expect(isValid).toBe(false);
        expect(wrapper.find('.error-text').text()).toEqual('Security level inclusion cannot be "Exclude Any". Note that the default value is now "Empty".');
    });
});
