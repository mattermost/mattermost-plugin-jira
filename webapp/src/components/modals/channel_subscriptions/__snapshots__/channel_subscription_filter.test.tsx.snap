// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ChannelSubscriptionFilter should match snapshot 1`] = `
<div
  className="row"
>
  <div
    className="col-md-12 col-sm-12"
  >
    <div
      className="help-text error-text"
      style={
        Object {
          "margin": "0 0 10px",
        }
      }
    />
  </div>
  <div
    className="col-md-11 col-sm-12"
  >
    <div
      className="row"
    >
      <div
        className="col-md-4 col-sm-12"
      >
        <ReactSelectSetting
          addValidate={
            [MockFunction] {
              "calls": Array [
                Array [
                  [Function],
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
          hideRequiredStar={true}
          name="fieldtype"
          onChange={[Function]}
          options={
            Array [
              Object {
                "label": "Affects versions",
                "value": "versions",
              },
              Object {
                "label": "Epic Link",
                "value": "customfield_10014",
              },
              Object {
                "label": "Fix versions",
                "value": "fixVersions",
              },
              Object {
                "label": "Labels",
                "value": "labels",
              },
              Object {
                "label": "MJK - Checkbox",
                "value": "customfield_10068",
              },
              Object {
                "label": "MJK - Labels",
                "value": "customfield_10071",
              },
              Object {
                "label": "MJK - Radio Buttons",
                "value": "customfield_10073",
              },
              Object {
                "label": "MJK - Select List 2",
                "value": "customfield_10075",
              },
              Object {
                "label": "MJK - Select List 3",
                "value": "customfield_10076",
              },
              Object {
                "label": "Priority",
                "value": "priority",
              },
              Object {
                "label": "Status",
                "value": "status",
              },
            ]
          }
          removeValidate={[MockFunction]}
          required={true}
          theme={Object {}}
          value={
            Object {
              "label": "Priority",
              "value": "priority",
            }
          }
        />
      </div>
      <div
        className="col-md-4 col-sm-12"
      >
        <ReactSelectSetting
          addValidate={
            [MockFunction] {
              "calls": Array [
                Array [
                  [Function],
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
          formatOptionLabel={[Function]}
          hideRequiredStar={true}
          name="inclusion"
          onChange={[Function]}
          options={
            Array [
              Object {
                "label": "Include",
                "value": "include_any",
              },
              Object {
                "label": "Exclude",
                "value": "exclude_any",
              },
              Object {
                "label": "Empty",
                "value": "empty",
              },
              Object {
                "label": "Include or Empty",
                "value": "include_or_empty",
              },
            ]
          }
          removeValidate={[MockFunction]}
          required={true}
          theme={Object {}}
          value={
            Object {
              "label": "Include",
              "value": "include_any",
            }
          }
        />
      </div>
      <div
        className="col-md-4 col-sm-12"
      >
        <ReactSelectSetting
          addValidate={
            [MockFunction] {
              "calls": Array [
                Array [
                  [Function],
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
          allowUserDefinedValue={false}
          hideRequiredStar={true}
          instanceID="https://something.atlassian.net"
          isClearable={false}
          isDisabled={false}
          isMulti={true}
          name="values"
          onChange={[Function]}
          options={
            Array [
              Object {
                "label": "Highest",
                "value": "1",
              },
              Object {
                "label": "High",
                "value": "2",
              },
              Object {
                "label": "Medium",
                "value": "3",
              },
              Object {
                "label": "Low",
                "value": "4",
              },
              Object {
                "label": "Lowest",
                "value": "5",
              },
            ]
          }
          removeValidate={[MockFunction]}
          required={true}
          resetInvalidOnChange={true}
          theme={Object {}}
          value={
            Array [
              Object {
                "label": "Highest",
                "value": "1",
              },
            ]
          }
        />
      </div>
    </div>
  </div>
  <div
    className="col-md-1 col-sm-12 text-center"
  >
    <button
      className="style--none"
      onClick={[Function]}
      style={
        Object {
          "color": undefined,
          "fontSize": "20px",
          "margin": "0.5rem 0 0",
        }
      }
      type="button"
    >
      <i
        className="fa fa-trash"
      />
    </button>
  </div>
  <ConfirmModal
    cancelButtonText="Cancel"
    confirmButtonClass="btn btn-danger"
    confirmButtonText="Delete"
    hideCancel={false}
    message="Are you sure you want to delete this filter?"
    onCancel={[Function]}
    onConfirm={[Function]}
    show={false}
    title="Field Filter"
  />
</div>
`;
